### DISCUSSION TITLE
edx编辑器和xmake适配遇到的一些问题

### DISCUSSION THREAD

**Discussion by heheda123123** [xmake-io/xmake#4266] [Ideas]
### 你在什么场景下需要该功能？

edx（edx介绍 https://zhuanlan.zhihu.com/p/635336347 ）和xmake的适配
![image](https://github.com/xmake-io/xmake/assets/97490782/328876df-e262-4662-90a0-94b5579c5a05)

1 有命令能关闭“编译输出的格式化”和“高亮”吗？
https://gitee.com/scu319hy/edx/issues/I76FIV#note_18886685_link

 2 添加include路径时，自动去掉了尾部的"/"，能保留吗
如图提到了可能会引发的问题

3 支持导出标准的数据格式
https://gitee.com/scu319hy/edx/issues/I76FIV#note_18892725_link
edx和clion的都采用了codeblocks的工程做为数据交换格式

### 描述可能的解决方案

1 给xmake加个命令指定输出原始的编译信息

[a comment by an automated account is not shown]

**Comment by heheda123123** [comment]
一个典型的codeblocks项目文件
[testwx31.txt](https://github.com/xmake-io/xmake/files/11757912/testwx31.txt)

[a comment by an automated account is not shown]

**Comment by waruqi** [comment]
> 1 有命令能关闭“编译输出的格式化”和“高亮”吗？

这个高亮是编译器输出的，xmake 只是设置了 -fdiagnostics-color=always 开启了而已

按理，如果你运行 xmake 时候捕获了 stdout，xmake 会自动关闭这个的，不会输出高亮，不需要额外的关闭，内部有 isatty 判断

如果非要强行关，可以运行时候配置 `XMAKE_COLORTERM=nocolor` 环境变量。

输出警告，加 -w 肯定有效果，完整输出不处理，走 -v

> 2 添加include路径时，自动去掉了尾部的"/"，能保留吗
如图提到了可能会引发的问题

不能保留

> 3 支持导出标准的数据格式

不支持 codeblocks 格式输出，但是可以执行 `xmake l xxx.lua`自定义脚本，实现任意格式输出，参考。

https://github.com/xmake-io/xmake-vscode/blob/master/assets/target_informations.lua

```lua
xmake l target_informations.lua
```

可以在这基础上，输出自己关心的数据，也可以自己输出符合 codeblocks 格式的数据。。


更多跟 ide 的交互，错误获取等，可以参考 xmake-vscode 插件的实现，大部分功能，这里面都有类似实现。

[a comment by an automated account is not shown]

**Comment by scu319hy** [comment]
我是EDX的作者

1. 关于高亮和格式化
相同的代码在cmake下输出是这样的
![image](https://github.com/xmake-io/xmake/assets/17721264/82c89ed9-08df-4bd7-9894-cbeadae2b0e0)
在xmake下是这样的
![image](https://github.com/xmake-io/xmake/assets/17721264/7323598a-e4a5-4468-9ca7-6bb44e1b7ea9)
可以看到，xmake输出是做了格式化的，文件路径变为相对路径了。并且在前面加了高亮的"error:"。cmake使用了-fdiagnostics-color=always，但并没有这种效果。

2. 关于警告加-w的不显示的问题。msvc下
同样的代码在cmake是这样
![image](https://github.com/xmake-io/xmake/assets/17721264/197f90ca-f3d8-4391-8632-a8b5cb9b7501)
而在xmake下是这样
![image](https://github.com/xmake-io/xmake/assets/17721264/b2c38005-e777-4c14-97a9-ddb38577fa56)
当然，gcc下加-w是有效果的。

3. 关于工程/目标信息生成的问题。
建议能官方支持一种标准的交换数据格式。方便其它工具集成。

通过脚本插件的形式去捕获这些数据最大的弊端是api接口缺少规范性。
我必须维护一段与xmake强依赖的代码。并且在远程编译时，我还要把这段代码传到服务器上去...
如果是用相对标准的交换数据文件，就能减少这种依赖性。
就算xmake版本变化，它也能够自成体系。而不是我再去根据它的版本变化来维护不同的插件。

[a comment by an automated account is not shown]

**Comment by waruqi** [comment]
> 可以看到，xmake输出是做了格式化的，文件路径变为相对路径了。并且在前面加了高亮的"error:"。cmake使用了-fdiagnostics-color=always，但并没有这种效果。

除了最开头那个 error: 前缀，后面是编译器的输出。。这边所有错误，不仅仅是编译错误，只要运行出错，执行到了 raise("xxx")

都会显示 `error: xxx` 

这个前缀是改不了的，但是高亮问题，上面说了 用 `XMAKE_COLORTERM=nocolor` 去运行 xmake 就能禁用高亮

参考：https://github.com/xmake-io/xmake-vscode/blob/c02f7c188dd9ebcabbd67a30507eee3d1f35d883/src/xmake.ts#L145

> 关于警告加-w的不显示的问题。msvc下

我这里可以，确保配置了 `set_warnings("all")` 对 msvc 开启所有警告输出

![image](https://github.com/xmake-io/xmake/assets/151335/6ada8fd6-a944-47da-adfc-76a5f5456827)

如果还是没有，可以看下 `xmake -v` 里面的完整编译参数，复制出来，手动执行下 看看是否有警告输出，或者跟 cmake 对下编译 flags 

> 关于工程/目标信息生成的问题。
建议能官方支持一种标准的交换数据格式。方便其它工具集成。
> 通过脚本插件的形式去捕获这些数据最大的弊端是api接口缺少规范性。
我必须维护一段与xmake强依赖的代码。并且在远程编译时，我还要把这段代码传到服务器上去...
如果是用相对标准的交换数据文件，就能减少这种依赖性。
就算xmake版本变化，它也能够自成体系。而不是我再去根据它的版本变化来维护不同的插件。

可以考虑内置，但目前暂时没时间支持，并且对这个格式 也不是很了解。

或者你可以先参考我上面说的 `xmake l codeblocks.lua` 方式，输出 codeblocks 格式输出，调通后，直接提 pr 过来到这个目录下 https://github.com/xmake-io/xmake/tree/master/xmake/plugins/project 就行了。

后面就能改用 `xmake project -k codeblocks` 执行。

[a comment by an automated account is not shown]
